Fix: Use Get-Volume for Windows snapshot size verification

shivamwayal37 opened the pull request:
#### What type of PR is this?

/kind failing-test

#### What this PR does / why we need it:

This PR fixes a long-standing issue where the filesystem size verification test was skipped or failing on Windows nodes due to platform incompatibility.

The e2e test `"should provision correct filesystem size when restoring snapshot to larger size pvc [Feature:VolumeSnapshotDataSource]"` relies on a utility function to determine volume capacity. The original logic was incompatible with Windows containers.

This PR implements robust, cross-platform logic by:

1.  **Refactoring:** Splitting the core logic into dedicated helpers: `getFilesystemSizeBytesWindows` and `getFilesystemSizeBytesLinux`.
2.  **Windows Fix:** Switching from the unreliable `fsutil` utility to the **PowerShell cmdlet `Get-Volume`**. This method executes `(Get-Volume -DriveLetter <X>).Size` to reliably retrieve a clean, numeric output for the total filesystem size, ensuring stability across diverse Windows CI environments (CAPZ, etc.).
3.  **Cleanup:** Addressing the previous logic's fragility and adopting the requested structural cleanup.

This change enables the full functionality of this important storage e2e test within the Windows CI jobs.

#### Which issue(s) this PR is related to:

Fixes #113359

#### Special notes for your reviewer:

/sig storage

This change includes the requested **refactoring** to split the original function into `getFilesystemSizeBytesWindows` and `getFilesystemSizeBytesLinux`. The Windows implementation now uses **PowerShell's `Get-Volume`** to avoid parsing issues and potential unavailability of `fsutil`, which was the likely cause of the persistent CAPZ E2E failures.

Local unit and integration verification passed successfully.

#### Does this PR introduce a user-facing change?
```release-note
NONE

shivamwayal37 committed:
Fix: Use Get-Volume for Windows snapshot size verification

Implement the robust, multi-stage PowerShell Get-Volume logic to accurately
determine the filesystem size of volumes mounted in Windows E2E tests. This
solution replaces the unreliable fsutil utility and resolves persistent
failures (including in CAPZ) caused by fragile parsing and directory mounts.

shivamwayal37 commented:
/check-milestone

shivamwayal37 commented:
/release-note-none

jsafrane commented:
Picking random Windows jobs
/test pull-kubernetes-e2e-capz-windows-master
/test pull-kubernetes-e2e-capz-azure-disk-windows
/test pull-kubernetes-e2e-capz-windows-serial-slow

jsafrane commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -1358,8 +1355,60 @@ func findVolumeMountPath(pod *v1.Pod, claim *v1.PersistentVolumeClaim) string {
+			return 0, fmt.Errorf("parsing total bytes: %w", err)
+		}
+
+		return totalBytes, nil
```
Please split it into `getFilesystemSizeBytesWindows` and `getFilesystemSizeBytesLinux` to keep the functions small and focused. `getFilesystemSizeBytes` will just have something like this

```go

	if framework.NodeOSDistroIs("windows") {
        return getFilesystemSizeBytesWindows(...)
    }
    return getFilesystemSizeBytesLinux(...)
```

shivamwayal37 commented:
/test pull-kubernetes-cmd

shivamwayal37 commented:
/test pull-kubernetes-e2e-capz-azure-disk-windows
/test pull-kubernetes-e2e-capz-windows-serial-slow

andyzhangx commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -1382,7 +1390,45 @@ func getFilesystemSizeBytes(pod *v1.Pod, mountPath string) (int, error) {
+	driveLetter := mountPath[:1] // e.g., "C"
+
+	// Construct PowerShell Command: Gets the numeric Size property directly.
+	psCmd := fmt.Sprintf("(Get-Volume -DriveLetter %s).Size", driveLetter)
```
I think this only works when the mountPath is on another drive, if mountPath is on c:\xx\xx, then it would get a wrong drive letter

shivamwayal37 replied:
Thanks, this is super helpful! 

Just to confirm the universal approach:

For **NTFS directory mounts** (e.g., `C:\k8s\mnt\volume`), the workflow is:

1.  Resolve underlying volume: `(Get-Item -Path "<mountPath>").Target`
2.  Convert volume ID to disk number: `(Get-Volume -UniqueId "<volumeID>" | Get-Partition).DiskNumber`
3.  Get disk size: `(Get-Disk -Number <diskNumber>).Size`

**My Clarification:** Is this **UniqueId → DiskNumber → Size** sequence recommended for *all* Kubernetes volume mounts on Windows, even when they are mounted as a dedicated drive (like `D:\`)?

Or, should we only use this complex sequence when the mount path starts with `C:\`?

Thanks again for the detailed guidance!

shivamwayal37 replied:
To fully align with CSI Proxy’s behavior, should I mirror the logic in `GetClosestVolumeIDFromTargetPath()` (resolving symlinks and walking up the hierarchy), or is directly calling `(Get-Item -Path $mountpath).Target` sufficient for all Windows E2E test environments?

shivamwayal37 commented:
Would `(Get-Volume -FilePath <mountPath>).Size` be the correct universal way to retrieve the filesystem size for both drive-letter mounts and mounted folders on Windows?

shivamwayal37 commented:
/test pull-kubernetes-e2e-capz-windows-master
/test pull-kubernetes-e2e-capz-azure-disk-windows
/test pull-kubernetes-e2e-capz-windows-serial-slow

shivamwayal37 commented:
/test pull-kubernetes-e2e-gce

shivamwayal37 commented:
## Windows Volume Resolution Refactor 

This update significantly refines how volume paths are resolved and managed on Windows, moving away from assumptions and aligning with CSI Proxy's robust logic.

---

### 1. Replaced Drive-Letter Assumption with Proper Volume Resolution

* **Old Logic:** Assumed all Windows mount paths began with a drive letter (e.g., `X:\`).
* **New Logic:** Resolves the **real underlying volume** using the following PowerShell command:
    ```powershell
    (Get-Item -Path "<mountPath>").Target
    ```
* **Benefit:** This approach correctly handles **drive letters**, **NTFS mount points**, and **symlinked paths**, ensuring broader compatibility.

---

### 2. Correct Handling for Both Volume Types

The logic now branches based on the resolved target type for accurate volume metadata retrieval:

* **If the target is a drive letter (e.g., `C`):**
    * Use `Get-Volume -DriveLetter <X>`
* **If the target is a Volume GUID (Globally Unique Identifier):**
    * Follows CSI Proxy's method to retrieve the necessary disk information:
        1.  `Get-Volume -UniqueId <guid>`
        2.  `Get-Partition` → resolve **DiskNumber**
        3.  `Get-Disk -Number <n>` → retrieve **Size**

---

### 3. Added Shared PowerShell Execution Helper

A new helper function was introduced to manage PowerShell execution consistently:

* Parses output **safely** (trims null bytes and whitespace).
* Provides **consistent error reporting**.
* Facilitates **easier future maintenance** of PowerShell logic.

---

### 4. Fully Aligned with CSI Proxy Logic

* **Mirrors** the exact logic CSI Proxy uses to resolve **volume IDs** and **disk sizes**.
* **Ensures** End-to-End (E2E) tests now behave **exactly like CSI drivers** on Windows nodes, validating real-world scenarios.

---

### 5. Fixes Real Cluster Scenarios

This refactor eliminates several previously reported issues and enhances compatibility:

* **Correct for NTFS directory mounts** (e.g., a volume mounted as a subdirectory on `C:\`).
* **Correct for GUID-backed PVCs** (Persistent Volume Claims).
* **Works on CAPZ clusters** (Cluster API Provider Azure) even when **no drive letters** are used.
* **Eliminates incorrect size reporting** (specifically fixing the issue seen in `#113359`).

---

### 6. No Change to Linux Paths / Test Behavior

* Only **Windows detection** logic has been improved.
* Tests on all platforms still expect the same **int64 byte-size** output as before.

shivamwayal37 commented:
## Summary of Windows Volume Resolution Refactor

### Key Improvements

1. **Universal Volume Resolution**  
   The logic no longer relies on fragile drive-letter assumptions (`X:\`). It now uses PowerShell’s `(Get-Item -Path ...).Target` to resolve the actual underlying volume for:
   - Drive letters  
   - NTFS directory mounts  
   - Symlinks  

2. **CSI Proxy Alignment**  
   Volume size retrieval now mirrors the CSI Proxy’s robust approach, using branching logic to correctly handle both simple drive letters and complex Volume GUIDs.

3. **Accurate Size Retrieval**  
   For GUID-based volumes, the function now follows the correct resolution chain:  
   `Get-Volume -UniqueId → Get-Partition → Get-Disk`,  
   ensuring accurate total disk size reporting.

4. **Improved Maintainability**  
   A shared PowerShell execution helper centralizes command execution, ensures safe and consistent output parsing (e.g., trimming null bytes/whitespace), and provides clearer error diagnostics.

5. **Fixes Critical E2E Failures**  
   The refactor corrects size reporting for volumes mounted as NTFS directories—resolving bug [#113359](https://github.com/kubernetes/kubernetes/issues/113359)—and improves reliability across CAPZ clusters.

andyzhangx reviewed (commented):
the code logic lgtm, does this PR passes on windows e2e test?

shivamwayal37 commented:
Yes — the required Windows tests are passing.

The jobs that failed (`capz-windows-serial-slow` and `capz-azure-disk-windows`) have not been passing since October, as mentioned earlier in conversation with @jsafrane. They’re known-flaky and non-blocking, and their failures are unrelated to the volume resolution logic in this PR.

The main Windows signal job (`pull-kubernetes-e2e-capz-windows-master`) passed, so the PR is safe from a Windows perspective.

andyzhangx reviewed (approved):
/lgtm

shivamwayal37 commented:
Thanks, @andyzhangx! Appreciate the review and approval.

shivamwayal37 commented:
Hi @jingxu97, @pohly, @carlory —  
This PR now has /lgtm and the required Windows signal job has passed.  
When you have a moment, could you please take a look and add /approve if it looks good?

Thank you!

shivamwayal37 commented:
Hi @xing-yang — just a gentle follow-up on the question about `VolumeSnapshotDataSource` being GA and whether the feature label can be removed so the test runs on Windows.

This PR already has /lgtm and all required Windows jobs are passing.

Happy to help with a follow-up PR if needed.

Thanks!
